php/php.code.analysis/src/org/netbeans/modules/php/analysis/PHPStanAnalyzerImpl.java

```diff
@@ -0,0 +1,249 @@
+import org.netbeans.api.annotations.common.CheckForNull;
+import org.netbeans.api.annotations.common.StaticResource;
+import org.netbeans.api.fileinfo.NonRecursiveFolder;
+import org.netbeans.modules.analysis.spi.Analyzer;
+import org.netbeans.modules.php.analysis.commands.PHPStan;
+import org.netbeans.modules.php.analysis.options.AnalysisOptions;
+import org.netbeans.modules.php.analysis.ui.analyzer.PHPStanCustomizerPanel;
+import org.netbeans.modules.php.analysis.util.AnalysisUtils;
+import org.netbeans.modules.php.analysis.util.Mappers;
+import org.netbeans.modules.php.api.executable.InvalidPhpExecutableException;
+import org.netbeans.modules.php.api.util.StringUtils;
+import org.netbeans.modules.refactoring.api.Scope;
+import org.netbeans.spi.editor.hints.ErrorDescription;
+import org.netbeans.spi.editor.hints.HintsController;
+import org.openide.filesystems.FileObject;
+import org.openide.filesystems.FileUtil;
+import org.openide.util.NbBundle;
+import org.openide.util.lookup.ServiceProvider;
+
+public class PHPStanAnalyzerImpl implements Analyzer {
+
+    private static final Logger LOGGER = Logger.getLogger(PHPStanAnalyzerImpl.class.getName());
+    private final Context context;
+    private final AtomicBoolean cancelled = new AtomicBoolean();
+
+    public PHPStanAnalyzerImpl(Context context) {
+        this.context = context;
+    }
+
+    @NbBundle.Messages({
+        "PHPStanAnalyzerImpl.phpStan.error=PHPStan is not valid",
+        "PHPStanAnalyzerImpl.phpStan.error.description=Invalid phpstan set in IDE Options."
+    })
+    @Override
+    public Iterable<? extends ErrorDescription> analyze() {
+        Preferences settings = context.getSettings();
+        if (settings != null && !settings.getBoolean(PHPStanCustomizerPanel.ENABLED, false)) {
+            return Collections.emptyList();
+        }
+
+        PHPStan phpStan = getValidPHPStan();
+        if (phpStan == null) {
+            context.reportAnalysisProblem(
+                    Bundle.PHPStanAnalyzerImpl_phpStan_error(),
+                    Bundle.PHPStanAnalyzerImpl_phpStan_error_description());
+            return Collections.emptyList();
+        }
+
+        String level = getValidPHPStanLevel();
+        FileObject config = getValidPHPStanConfiguration();
+        PHPStanParams phpStanParams = new PHPStanParams()
+                .setLevel(level)
+                .setConfiguration(config);
+        Scope scope = context.getScope();
+
+        Map<FileObject, Integer> fileCount = AnalysisUtils.countPhpFiles(scope);
+        int totalCount = 0;
+        for (Integer count : fileCount.values()) {
+            totalCount += count;
+        }
+
+        context.start(totalCount);
+        try {
+            return doAnalyze(scope, phpStan, phpStanParams, fileCount);
+        } finally {
+            context.finish();
+        }
+    }
+
+    @Override
+    public boolean cancel() {
+        cancelled.set(true);
+        // XXX cancel phpstan ?
```

Comment:
Maybe another issue/question for PHPStan?

Reply:
Just a nitpick, of course, nothing important.

Reply:
Code Sniffer also has this comment. Actually, `phpstan` command is not canceled here. (If there is the next target, the command is not run for it.) However, users can stop the command in the output window. So, I think we can remove this comment :)